### PR TITLE
deprecate ERR_GET_FUNC()

### LINKED ISSUE #15946: ERR_GET_FUNC should be removed or marked deprecated
Handling of encrypted private keys in `mod_ssl` was broken by the change to remove function codes, since it checked for a specific function returning errors in:

https://github.com/apache/httpd/blob/trunk/modules/ssl/ssl_engine_init.c#L1484

IMO if function codes are removed it would be preferable to remove ERR_GET_FUNC completely - the ABI is changed here, better to have it break at compile time than at runtime in an obscure way.  Or at least ERR_GET_FUNC should be marked deprecated so we get a gcc warning, which it isn't currently?

### PR DESCRIPTION
**Pull request by paulidale**
Deprecate the ERR_GET_FUNC() function.
Change it's return to -1 so it can't accidentally match one of the ERR_F_ macros.
Update documentation to note the deprecation.
Update documentation so it doesn't claim the ERR_GET_ family are macros.

- [x] documentation is added or updated
- [ ] tests are added or updated

Fixes #15946

### CODE REVIEW

**Comment by paulidale** [comment]
@reaperhulk, do the Python tests depend on the function codes?  I can't see anything but I'm not sure how else the CI failure is relevant.

**Comment by reaperhulk** [comment]
@paulidale Yeah this test asserts on a func code. We've been progressively making our error stack checks less specific to be more resilient but it looks like we'd be best served entirely stripping `ERR_GET_FUNC`.

I can update `cryptography`'s main branch tomorrow and send a PR to openssl bumping to latest, which should unblock this.

**Comment by paulidale** [comment]
Thanks, that would be perfect.

**Comment by slontis** [comment]
Shouldnt this have the OTC hold on it also - since the issue has a hold.

**Review by levitte — APPROVED** [review]

**Review by slontis — CHANGES_REQUESTED** [review]
Add to the 'Deprecated list' in the migration_guide.pod (Somewhere around line 1610)

**Comment by paulidale** [comment]
Added an entry to the migration guide

**Comment by slontis** [comment]
LGTM: This should have a OTC hold on it right?

**Comment by paulidale** [comment]
The issue has one, might as well put one here too.

**Review by t8m — APPROVED** [review]
LGTM to me as well. IMO this is way better than a silent breakage that might happen in applications with what we currently have.

**Comment by t8m** [comment]
How do we resolve the pyca test breakage though? Maybe temporarily disable the test and wait for the upstream to merge pyca/cryptography#6148

**Comment by paulidale** [comment]
The pyca change should be in first and then we'll get a [PR to merge](https://github.com/openssl/openssl/pull/15955#issuecomment-871083378) it for our external tests.  This goes in afterwards or we live with it breaking for a short time -- I'd prefer the former.

**Comment by romen** [comment]
Uhm, wouldn't it be better to just remove `ERR_GET_FUNC()` completely?

Failures at compile time are more desirable than at run time for this, because of the likelyhood of this function being used mostly in branches never/seldom taken.

I tend to believe that keeping the function (as deprecated) as proposed, but subtly changing its behavior, is not really helping external authors relying on our promise of API stability for 3.0 anyway, that is why I am considering direct removal.

Regarding doing this now, I believe the underlying issue is indeed an API bug, which would warrant an exemption from the beta API freeze, and should definitely go in before final release. (i.e., I would vote +1 in an OTC vote to fix this issue by changing this part of the API at this time)

**Comment by t8m** [comment]
I agree with @romen that removing the call completely would be also an option and maybe the best one.

**Comment by mattcaswell** [comment]
OTC have voted for removal

**Comment by paulidale** [comment]
Closing this, it is not the path forwards.

**Pull request closed without merging**
